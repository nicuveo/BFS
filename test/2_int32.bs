def impure check(S op) {
    >>>>>>>>>>>>
    [[-] prints("ERROR: garbage remaining in +12 after ") prints(op) prints("\n")] <
    [[-] prints("ERROR: garbage remaining in +11 after ") prints(op) prints("\n")] <
    [[-] prints("ERROR: garbage remaining in +10 after ") prints(op) prints("\n")] <
    [[-] prints("ERROR: garbage remaining in +9 after ")  prints(op) prints("\n")] <
    [[-] prints("ERROR: garbage remaining in +8 after ")  prints(op) prints("\n")] <
    [[-] prints("ERROR: garbage remaining in +7 after ")  prints(op) prints("\n")] <
    [[-] prints("ERROR: garbage remaining in +6 after ")  prints(op) prints("\n")] <
    [[-] prints("ERROR: garbage remaining in +5 after ")  prints(op) prints("\n")] <
    [[-] prints("ERROR: garbage remaining in +4 after ")  prints(op) prints("\n")] <
    [[-] prints("ERROR: garbage remaining in +3 after ")  prints(op) prints("\n")] <
    [[-] prints("ERROR: garbage remaining in +2 after ")  prints(op) prints("\n")] <
    [[-] prints("ERROR: garbage remaining in +1 after ")  prints(op) prints("\n")] <
}

def test_print(I x) {
    pushi(x)
    printi check("printi")
    prints("\n")
    popi
}

def test_add(I x, I y) {
    pushi(x) printi prints(" + ")
    pushi(y) printi prints(" = ")
    addi check("addi")
    printi prints("\n")
    popi
}

def test_sub(I x, I y) {
    pushi(x) printi prints(" - ")
    pushi(y) printi prints(" = ")
    swapi
    subi check("subi")
    printi prints("\n")
    popi
}

def main() {
    // // printing
    test_print(0x80000000)
    test_print(0xFFFFFFFF)
    test_print(0x7FFFFFFF)

    // adding
    test_add(0x0000FFFF, 1)
    test_add(0x00FFFFFF, 0x00FFFFFF)

    // adding
    test_sub(10, 20)
    test_sub(0x1000, 1)
    test_sub(-42, -84)
    test_sub(0xFFFFF000, -1)
    test_sub( 42, -10)
    test_sub(-10,  42)
}
