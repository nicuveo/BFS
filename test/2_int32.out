-2147483648
-1
2147483647
65535 + 1 = 65536
16777215 + 16777215 = 33554430
10 - 20 = -10
4096 - 1 = 4095
-42 - -84 = 42
-4096 - -1 = -4095
42 - -10 = 52
-10 - 42 = -52
